src/ngraph_translateops.h

```diff
@@ -0,0 +1,269 @@
+ *******************************************************************************/
+
+#include "ngraph_conversions.h"
+#include "ngraph_utils.h"
+
+using namespace std;
+namespace ng = ngraph;
+
+namespace tensorflow {
+
+namespace ngraph_bridge {
+
+using VectNg = std::vector<shared_ptr<ng::Node>>;
+
+// TODO (sarkars): combine the 2 MakePaddings together? with default args?
+// would want the default args to be at the end of the paramlist,
+// but 'outputs' (like ng_padding_below, ng_padding_above) are usually at the
+// end of the param list
+// TODO (sarkars): why does MakePadding handle only 2 dimensions... generalize
+// it?
+// TODO (sarkars): add unit tests for 1D conv. 1d pooling etc. check if
+// MakePadding works in that case
+template <typename T>
+void MakePadding(const std::string& tf_padding_type,
+                 const ngraph::Shape& ng_image_shape,
+                 const ngraph::Shape& ng_kernel_shape,
+                 const ngraph::Strides& ng_strides,
+                 const ngraph::Shape& ng_dilations, T& ng_padding_below,
+                 T& ng_padding_above) {
+  ngraph::Shape ng_dilation_kernel_shape{
+      (ng_kernel_shape[0] - 1) * ng_dilations[0] + 1,
+      (ng_kernel_shape[1] - 1) * ng_dilations[1] + 1};
+
+  MakePadding(tf_padding_type, ng_image_shape, ng_dilation_kernel_shape,
+              ng_strides, ng_padding_below, ng_padding_above);
+}
+
+template <typename T>
+void MakePadding(const std::string& tf_padding_type,
+                 const ngraph::Shape& ng_image_shape,
+                 const ngraph::Shape& ng_kernel_shape,
+                 const ngraph::Strides& ng_strides, T& ng_padding_below,
+                 T& ng_padding_above) {
+  if (tf_padding_type == "SAME") {
+    for (size_t i = 0; i < 2; i++) {
+      /*
+      size_t image_size = ng_image_shape[i];
+      size_t filter_shape = ng_kernel_shape[i];
+      size_t filter_stride = ng_strides[i];
+
+      int64 padding_needed;
+      if (image_size % filter_stride == 0) {
+        padding_needed = filter_shape - filter_stride;
+      } else {
+        padding_needed = filter_shape - (image_size % filter_stride);
+      }
+      if (padding_needed < 0) {
+        padding_needed = 0;
+      }
+      */
+
+      // TODO: check this:. This formula is documented well documented here.
+      // So I prefer this, compared to the one above (though both are exactly
+      // the same)
+      // https://www.tensorflow.org/api_guides/python/nn#Notes_on_SAME_Convolution_Padding
+      int64 out_shape = ceil(ng_image_shape[i] / ng_strides[i]);
+      int64 padding_needed = ng_strides[i] * (out_shape - 1) +
+                             ng_kernel_shape[i] - ng_image_shape[i];
+      padding_needed = padding_needed < 0 ? 0 : padding_needed;
+
+      size_t padding_lhs = padding_needed / 2;
+      size_t padding_rhs = padding_needed - padding_lhs;
+      ng_padding_below[i] = padding_lhs;
+      ng_padding_above[i] = padding_rhs;
```

Comment:
I have a dim recollection that I may have come across that document before, taken the formula as it was, and found out that it wasn't quite correct. If I'm wrong, let's definitely use the version from the docs.

Reply:
the padding issue: I ran the math on both implementations, they seem to be the same. But who knows what crazy corner case might show up. In case the current implementation works, lets keep that

Reply:
added new commit keeping old formula